Add NetworkID to host metadata

leeavital opened the pull request:
### What does this PR do?

- Adds the concept of a network_id, a field that will be used to help with network address resolution to the agent
- The network ID is  optional, and detected as the VPC ID for both EC2 and GCE
- The network ID is reported as part of host metadata
- A follow up commit will include the network ID in the connections payload reported by the process-agent. Since both core and process agent need to be aware of this ID, the configuration option (network.id) is part of the core config package, and the detection code is in the shared util package.  
          

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

leeavital committed:
This will be helpful for resolving connection payloads.
The ID will eventually be piped through the core agent as host metadata
and through the process agent as container metadata.

leeavital committed:
Add known config

leeavital committed:
Add network ID to host payload

hush-hush commented on pkg/util/ec2/ec2_test.go:
```diff
@@ -107,3 +107,27 @@ func TestExtractClusterName(t *testing.T) {
 	}
 }
+
+func TestGetNetworkID(t *testing.T) {
```
Can we add test for the case where we have no networkID and multiple ones (ie the error cases) ?

hush-hush replied:
Can we add test for the case where we have no networkID and multiple ones (ie: the error cases) ?

hush-hush commented on pkg/util/ec2/ec2.go:
```diff
@@ -35,6 +36,34 @@ func GetHostname() (string, error) {
 	return getMetadataItemWithMaxLength("/hostname", config.Datadog.GetInt("metadata_endpoints_max_hostname_size"))
 }
 
+func GetNetworkID() (string, error) {
```
You need to add a comment to all exported function for the tests to pass.

hush-hush commented on pkg/util/gce/gce.go:
```diff
@@ -75,6 +77,35 @@ func GetClusterName() (string, error) {
 	return clusterName, nil
 }
 
+func GetNetworkID() (string, error) {
```
Same here.

hush-hush commented on pkg/util/ec2/ec2.go:
```diff
@@ -35,6 +36,34 @@ func GetHostname() (string, error) {
+	}
+
+	macs := strings.Split(strings.TrimSpace(resp), "\n")
+	vpcIDs := common.NewStringSet()
```
Not familiar with GCE and AWS networking: is having the same network ID for multiple `mac` a real use case ?

hush-hush commented on pkg/util/network.go:
```diff
@@ -0,0 +1,44 @@
+// * configuration
+// * GCE
+// * EC2
+func GetNetworkID() (string, error) {
```
Will this be used anywhere else than the host metadata provider ? If no we should move it there with the rest of the metadata logic.

leeavital replied:
I'm not sure what you're asking. I thought this *was* the host metadata provider.

I mentioned it briefly in the commit description, we're also going to be calling `GetNetworkID()` in the process agent so we can attach this to container payloads.  I'm not sure if that changes things.

hush-hush replied:
:+1:, just making sure this function belong in `pkg/util`.

hush-hush commented on pkg/util/network.go:
```diff
@@ -0,0 +1,44 @@
+	log.Debugf("GetNetworkID trying GCE")
+	if networkID, err := gce.GetNetworkID(); err != nil {
+		cache.Cache.Set(cacheNetworkIDKey, networkID, cache.NoExpiration)
+		return networkID, nil
```
Can we add another debug log line saying the the networkID was found with GCE (and even logging the networkID if it's not a sensitive info) ?

leeavital replied:
Yes, done.

hush-hush commented on pkg/util/network.go:
```diff
@@ -0,0 +1,44 @@
+	log.Debugf("GetNetworkID trying EC2")
+	if networkID, err := ec2.GetNetworkID(); err != nil {
+		cache.Cache.Set(cacheNetworkIDKey, networkID, cache.NoExpiration)
+		return networkID, nil
```
same here.

hush-hush commented on pkg/metadata/host/host.go:
```diff
@@ -145,6 +146,11 @@ func getMeta() *Meta {
 }
 
+func getNetworkMeta() *NetworkMeta {
+	nid, _ := util.GetNetworkID()
```
We should log the error from `GetNetworkID`. Also in case of error do we want to return a `NetworkMeta` with an empty string or return `nil` ?

hush-hush replied:
`getMeta` follows agent5 payload which use to send an empty JSON dict. We could do better here since network ID is new. It depends on the backend implementation.

leeavital replied:
I change this to return nil if the network meta is missing

hush-hush commented:
You also need a release note for this I think.

leeavital committed:
Review feedback:

- Update comment in ec2.go and gce.go
- Make tests more comprehensive
- Add Debug lines

leeavital committed:
Add release notes

leeavital committed:
Add a comment explaining NetworkMeta

hush-hush commented on pkg/util/ec2/ec2_test.go:
```diff
@@ -107,3 +107,66 @@ func TestExtractClusterName(t *testing.T) {
+	metadataURL = ts.URL
+
+	_, err := GetNetworkID()
+	assert.Error(t, err)
```
Could we also check for the error message here, making the right error is returned.

hush-hush commented on pkg/util/ec2/ec2_test.go:
```diff
@@ -107,3 +107,66 @@ func TestExtractClusterName(t *testing.T) {
+	metadataURL = ts.URL
+
+	_, err := GetNetworkID()
+	assert.Error(t, err)
```
Could we also check for the error message here, making the right error is returned.

hush-hush commented on pkg/util/gce/gce_test.go:
```diff
@@ -86,3 +86,62 @@ func TestGetClusterName(t *testing.T) {
+	metadataURL = ts.URL
+
+	_, err := GetNetworkID()
+	assert.Error(t, err)
```
Same here.

hush-hush commented on pkg/util/gce/gce_test.go:
```diff
@@ -86,3 +86,62 @@ func TestGetClusterName(t *testing.T) {
+	metadataURL = ts.URL
+
+	_, err := GetNetworkID()
+	assert.Error(t, err)
```
And here

hush-hush commented on releasenotes/notes/report-network-id-43e5afaa0976aeae.yaml:
```diff
@@ -0,0 +1,13 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+upgrade:
```
extra `upgrade` here.

leeavital committed:
Log error in getNetworkMeta

leeavital committed:
Too many addresses

leeavital committed:
Update tests

leeavital committed:
Fix release notes

leeavital committed:
fixup host.go

leeavital committed:
Use BindEnv, correctly handle errors

leeavital commented:
@hush-hush  feedback should be addressed.

I also changed the config from SetKnown to BindEnv to allow setting the configuration with an environment variable.
